r/R/filesystem.R

```diff
@@ -232,6 +243,17 @@ LocalFileSystem$create <- function() {
 #' @export
 SubTreeFileSystem <- R6Class("SubTreeFileSystem", inherit = FileSystem)
 SubTreeFileSystem$create <- function(base_path, base_fs) {
-  xp <- fs___SubTreeFileSystem__create(base_path, base_fs)
+  xp <- fs___SubTreeFileSystem__create(fs_relative_path(base_path), base_fs)
   shared_ptr(SubTreeFileSystem, xp)
 }
+
+clean_path <- function(path) {
+  # Make sure we have a valid, absolute, forward-slashed path for passing to Arrow
+  normalizePath(path, winslash = "/", mustWork = FALSE)
+}
+
+fs_relative_path <- function(path) {
+  # Make sure all path separators are "/", not "\" as on Windows
+  path_sep <- ifelse(tolower(Sys.info()[["sysname"]]) == "windows", "\\\\", "/")
```

Comment:
You should probably use `.Platform$file.sep`

Reply:
Also not sure about the name of this function, it is not really doing anything to compute a relative path, you could probably just use `normalizePath(mustWork = FALSE)`, assuming path doesn't actually have to exist on the filesystem.

Reply:
Re: `.Platform$file.sep`, I intended to but then on my Windows VM, `.Platform$file.sep` reported "/" while `tempfile()` still created paths with "\"  🤷‍♂  

I originally did `normalizePath` but unfortunately for this use case, `normalizePath` does path expansion, even with `mustWork=FALSE`. We need to keep the path relative (to whatever the FileSystem object holds as its cwd) and just need Windows `\` to be converted to `/`. Regular expressions were my last resort and I'm happy to use something better or more robust if it exists.

Reply:
It looks a bit weird to do a no-op replacement on Unix, though I suppose it doesn't work.